src/controller/stream.rs

```diff
@@ -0,0 +1,341 @@
+use std::{future::Future, sync::Arc};
+
+use futures::FutureExt;
+use subsquid_messages::{data_chunk::DataChunk, query_result, Range};
+use tokio::task::JoinSet;
+use tracing::instrument;
+
+use crate::{
+    controller::timeouts::TimeoutManager,
+    network::NetworkClient,
+    types::{ClientRequest, DatasetId, RequestError, ResponseChunk},
+    utils::SlidingArray,
+};
+
+pub struct StreamController {
+    request: ClientRequest,
+    stream_id: usize,
+    network: Arc<NetworkClient>,
+    next_index: usize,
+    next_chunk: Option<DataChunk>,
+    downloads: JoinSet<(Result<ResponseChunk, RequestError>, usize)>,
+    buffer: SlidingArray<Option<Result<ResponseChunk, RequestError>>>,
+    timeouts: Arc<TimeoutManager>,
+    // The first request is sent out to probe it's validity.
+    // Only after that multiple requests are sent in parallel.
+    trial: bool,
```

Comment:
This strategy will slow down the streaming bootstrap, while saving just a little bit of money. Likely not worth it.

Reply:
Okay, maybe you're right. I thought only the throughput is important, not the latency